cpp/src/arrow/csv/converter.cc

```diff
@@ -191,6 +205,61 @@ class VarSizeBinaryConverter : public ConcreteConverter {
   }
 };
 
+template <typename T, bool CheckUTF8>
+class DictionaryBinaryConverter : public ConcreteDictionaryConverter {
+ public:
+  using ConcreteDictionaryConverter::ConcreteDictionaryConverter;
+
+  Status Convert(const BlockParser& parser, int32_t col_index,
+                 std::shared_ptr<Array>* out) override {
+    // We use a fixed index width so that all column chunks get the same index type
+    using BuilderType = Dictionary32Builder<T>;
+    BuilderType builder(type_, pool_);
+
+    auto visit_non_null = [&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
+      if (CheckUTF8 && ARROW_PREDICT_FALSE(!util::ValidateUTF8(data, size))) {
+        return Status::Invalid("CSV conversion error to ", type_->ToString(),
+                               ": invalid UTF8 data");
+      }
+      RETURN_NOT_OK(
+          builder.Append(util::string_view(reinterpret_cast<const char*>(data), size)));
+      if (ARROW_PREDICT_FALSE(builder.dictionary_length() > max_cardinality_)) {
+        return Status::IndexError("Dictionary length exceeded max cardinality");
+      }
+      return Status::OK();
+    };
+
+    RETURN_NOT_OK(builder.Resize(parser.num_rows()));
+
+    if (options_.strings_can_be_null) {
+      auto visit = [&](const uint8_t* data, uint32_t size, bool quoted) -> Status {
+        if (IsNull(data, size, false /* quoted */)) {
+          return builder.AppendNull();
+        } else {
+          return visit_non_null(data, size, quoted);
+        }
+      };
+      RETURN_NOT_OK(parser.VisitColumn(col_index, visit));
+    } else {
+      RETURN_NOT_OK(parser.VisitColumn(col_index, visit_non_null));
+    }
+
+    RETURN_NOT_OK(builder.Finish(out));
+
+    return Status::OK();
```

Comment:
`return build.Finish(out);`

Reply:
This is how it's done in the rest of the file, I'd rather keep the conventions consistent.